packages/experiments/src/components/PositioningContainer/PositioningContainer.tsx

```diff
@@ -0,0 +1,387 @@
+   */
+  positions?: IPositionInfo;
+
+  /**
+   * Tracks the current height offset and updates during
+   * the height animation when props.finalHeight is specified.
+   */
+  heightOffset?: number;
+}
+
+export class PositioningContainer extends BaseComponent<IPositioningContainerProps, IPositioningContainerState> {
+
+  public static defaultProps = {
+    preventDismissOnScroll: false,
+    offsetFromTarget: 0,
+    minPagePadding: 8,
+    directionalHint: DirectionalHint.bottomAutoEdge
+  };
+
+  private _didSetInitialFocus: boolean;
+
+  /**
+   * The primary positioned div.
+   */
+  private _positionedHost: HTMLDivElement;
+
+  // @TODO rename to reflect the name of this class
+  private _contentHost: HTMLDivElement;
+
+  /**
+   * Stores an instance of Window, used to check
+   * for server side rendering and if focus was lost.
+   */
+  private _targetWindow: Window;
+
+  /**
+   * The bounds used when determing if and where the
+   * PositioningContainer should be placed.
+   */
+  private _positioningBounds: IRectangle;
+
+  /**
+   * The maximum height the PositioningContainer can grow to
+   * without going being the window or target bounds
+   */
+  private _maxHeight: number | undefined;
+  private _positionAttempts: number;
+  private _target: HTMLElement | MouseEvent | IPoint | null;
+  private _setHeightOffsetTimer: number;
+
+  constructor(props: IPositioningContainerProps) {
+    super(props);
+    this._didSetInitialFocus = false;
+    this.state = {
+      positions: undefined,
+      heightOffset: 0
+    };
+    this._positionAttempts = 0;
+  }
+
+  public componentWillMount() {
+    this._setTargetWindowAndElement(this._getTarget());
+  }
+
+  public componentDidMount() {
+    this._onComponentDidMount();
+  }
+
+  public componentDidUpdate() {
+    this._setInitialFocus();
+    this._updateAsyncPosition();
+  }
+
+  public componentWillUpdate(newProps: IPositioningContainerProps) {
+    // If the target element changed, find the new one. If we are tracking
+    // target with class name, always find element because we do not know if
+    // fabric has rendered a new element and disposed the old element.
+    let newTarget = this._getTarget(newProps);
+    let oldTarget = this._getTarget();
+    if (newTarget !== oldTarget || typeof (newTarget) === 'string' || newTarget instanceof String) {
+      this._maxHeight = undefined;
+      this._setTargetWindowAndElement(newTarget!);
+    }
+
+    if (newProps.offsetFromTarget !== this.props.offsetFromTarget) {
+      this._maxHeight = undefined;
+    }
+
+    if (newProps.finalHeight !== this.props.finalHeight) {
+      this._setHeightOffsetEveryFrame();
+    }
+  }
+
+  public render() {
+    // If there is no target window then we are likely in server side rendering and we should not render anything.
+    if (!this._targetWindow) {
+      return null;
+    }
+
+    let {
+      className,
+      calloutWidth,
```

Comment:
Nit: Rename all "callout" stuff with positioningContainer or something to that effect.